plugins/woocommerce-blocks/assets/js/atomic/utils/register-block-single-product-template.ts

```diff
@@ -1,140 +1,319 @@
-			if (
-				! isRegistered &&
-				isAvailableOnPostEditor &&
-				! editSiteStore
-			) {
-				if ( isVariationBlock ) {
-					blockRegistrationAttempts.add( variationName );
-					registerBlockVariation(
-						blockName,
-						blockSettings as BlockVariation< BlockAttributes >
-					);
-				} else {
-					blockRegistrationAttempts.add( blockName );
-					// @ts-expect-error: `registerBlockType` is typed in WordPress core
-					registerBlockType( blockMetadata, blockSettings );
-				}
+	/**
+	 * Unregisters a block or block variation.
+	 * Handles both regular blocks and variations with error handling.
+	 *
+	 * @param {BlockConfig} config - Configuration of the block to unregister
+	 */
+	private unregisterBlock( config: BlockConfig ): void {
+		const { blockName, isVariationBlock, variationName } = config;
+
+		try {
+			if ( isVariationBlock && variationName ) {
+				unregisterBlockVariation( blockName, variationName );
+				this.registeredBlocks.delete( variationName );
+			} else {
+				unregisterBlockType( blockName );
+				this.registeredBlocks.delete( blockName );
 			}
-		}, 'core/edit-post' );
+		} catch ( error ) {
+			// eslint-disable-next-line no-console
+			console.debug(
+				`Failed to unregister block ${ blockName }:`,
+				error
+			);
+		}
+	}
+
+	/**
+	 * Registers a block or block variation.
+	 * Handles different registration requirements for various contexts.
+	 * Includes checks to prevent recursive registration.
+	 *
+	 * @param {BlockConfig} config - Configuration of the block to register
+	 */
+	private registerBlock( config: BlockConfig ): void {
+		const {
+			blockName,
+			blockMetadata = blockName,
+			blockSettings,
+			isVariationBlock,
+			variationName,
+			isAvailableOnPostEditor,
+		} = config;
+
+		if ( this.isRegistering ) {
+			return;
+		}
+
+		try {
+			this.isRegistering = true;
+
+			// Check if block is already registered
+			const key = variationName || blockName;
+			if ( this.registeredBlocks.has( key ) ) {
+				return;
+			}
+
+			const editSiteStore = select( 'core/edit-site' );
+
+			// Don't register if we're in post editor context and block isn't available there
+			if ( ! editSiteStore && ! isAvailableOnPostEditor ) {
+				return;
+			}
+
+			if ( isVariationBlock ) {
+				registerBlockVariation(
+					blockName,
+					blockSettings as BlockVariation< BlockAttributes >
+				);
+			} else {
+				const ancestor = isEmpty( blockSettings?.ancestor )
+					? [ 'woocommerce/single-product' ]
+					: blockSettings?.ancestor;
+
+				// Only remove ancestor if we're in site editor AND in single-product template
+				const shouldRemoveAncestor =
+					editSiteStore &&
+					this.currentTemplateId?.includes( 'single-product' );
+
+				// @ts-expect-error - blockMetadata can be either string or object
+				registerBlockType( blockMetadata, {
+					...blockSettings,
+					ancestor: shouldRemoveAncestor ? undefined : ancestor,
+				} );
+			}
+
+			this.registeredBlocks.add( key );
```

Comment:
I suggest to rename it to `blockRegistrationAttempts`. Technically, `registerBlockType` might not register the block. The `registeredBlocks` map is not reliable for checking if a block is registered or not. Improving the name ensures it won't be used for this kind of check.

Related discussion: https://github.com/woocommerce/woocommerce/pull/53217#issuecomment-2507922065

Reply:
Thanks for the context! Done in https://github.com/woocommerce/woocommerce/pull/53788/commits/cb2be812b7733f5a31c37af179dac19d60427bef